File: sample/gadt.hs
{-# LANGUAGE GADTs #-}
module GADT where

data TermInt = LitT Int | AddT TermInt TermInt

evalT :: TermInt -> Int
evalT (LitT n) = n
evalT (AddT t1 t2) = evalT t1 + evalT t2

data Term a where
    Lit  :: Int -> Term Int
    Add  :: Term Int -> Term Int -> Term Int
    Eql   :: Term Int -> Term Int -> Term Bool
    
deriving instance (Eq a) => Eq (Term a)
deriving instance (Show a) => Show (Term a)

eval :: Term a -> a
eval (Lit n) = n
eval (Add t1 t2) = eval t1 + eval t2
eval (Eql t1 t2) = eval t1 == eval t2

t1 = Add (Lit 1) (Add (Lit 2) (Lit 2))
t2 = Add (Lit 2) (Lit 4)

eq = eval (Eql t1 t2)

main :: IO ()
main = do
    print $ show $ evalT (AddT (LitT 2) (AddT (LitT 2) (LitT 3)))
    print $ show t1 ++ "is equal to " ++ show t2 ++ "? " ++ show eq

File: sample/gadt.kt
sealed interface Term<T> {
    fun eval(): T
}

data class Lit(val value: Int) : Term<Int> {
    override fun eval() = value
}

data class Add(val t1: Term<Int>, val t2: Term<Int>) : Term<Int> {
    override fun eval() = t1.eval() + t2.eval()
}

data class Eq(val t1: Term<Int>, val t2: Term<Int>): Term<Boolean> {
    override fun eval() = t1.eval() == t2.eval()
}

sealed interface IntTerm : Term<Int>
sealed interface BooleanTerm : Term<Boolean>

// fun IntTerm.evalExt(): Int = when (this) { 
//     is Lit -> value
//     is Add -> t1.evalExt() + t2.evalExt()
// }

// fun BooleanTerm.evalExt() = when(this) {
//     is Eq -> t1.evalExt() == t2.evalExt()
// }

// fun <T> eval(term: Term<T>) : T = 
//     when (term) {
//         is Lit -> term.value as Int
//         is Add -> eval(term.t1) as Int + eval(term.t2) as Int
//         is Eq -> eval(term.t1) as Int == eval(term.t2) as Int
//     } as T

fun main() {
    val term = Eq(Add(Lit(1), Lit(2)), Lit(3))
    // val result = eval(term)
    // val resultExt = term.evalExt()

    // println(result)
    // println(resultExt)

    println(term.eval())
}